Fix dev server serving inline bundles

martinsvoboda opened the pull request:
<!---
Thanks for filing a pull request 😄 ! Before you submit, please read the following:

Search open/closed issues before submitting since someone might have pushed the same thing before!
-->

Fixes #6043. Fix recommended by @mischnic 

Dev server fails when serving inline bundle processed by @parcel/transformer-raw

I would consider this as regression bug, because it worked with parcel@2.0.0-beta.2 and @parcel/transformer-raw@2.0.0-beta.2

<!---
Provide a general summary of the pull request here
Please look for any issues that this PR resolves and tag them in the PR.
-->

## 💻 Examples

<!-- Examples help us understand the requested feature better -->

package.json
```
{
    "name": "parcel-serve-fails-for-inlines",
    "scripts": {
        "serve": "parcel serve ./src/index.html"
    },
    "devDependencies": {
        "parcel": "2.0.0-nightly.632"
    }
}
```

.parcelrc
```
{
    "extends": "@parcel/config-default",
    "transformers": {
        "*.props": ["@parcel/transformer-raw"]
    },
}
```

index.html
```
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Title</title>
</head>
<body>
    <h1>Test</h1>
    <div data-widget="TextFromImage">
        <script type="text/props">
            {
                "some": "value"
            }
        </script>
    </div>
</body>
</html>
```


## 🚨 Test instructions

<!-- In case it is impossible (or too hard) to reliably test this feature/fix with unit tests, please provide test instructions! -->

I was unable to create the correct unit test. Reproducing the issue:

- Repo https://github.com/martinsvoboda/parcel-serve-fails-for-inlines
- Run `parcel serve ./src/index.html` it is finished successfully
- Open localhost:1234 in browser, serve process fails with following exception

```
parcel serve ./src/index.html

ℹ️ Server running at http://localhost:1234
✨ Built in 230ms
Error: Got unexpected null
    at nullthrows (/Users/martin.svoboda/workspace/parcel-serve-fails-for-inlines/node_modules/nullthrows/nullthrows.js:7:15)
    at PackagedBundle.get filePath [as filePath] (/Users/martin.svoboda/workspace/parcel-serve-fails-for-inlines/node_modules/@parcel/core/lib/public/Bundle.js:367:38)
    at /Users/martin.svoboda/workspace/parcel-serve-fails-for-inlines/node_modules/@parcel/reporter-dev-server/lib/Server.js:272:104
    at Array.find (<anonymous>)
    at Server.serveBundle (/Users/martin.svoboda/workspace/parcel-serve-fails-for-inlines/node_modules/@parcel/reporter-dev-server/lib/Server.js:272:45)
    at Server.respond (/Users/martin.svoboda/workspace/parcel-serve-fails-for-inlines/node_modules/@parcel/reporter-dev-server/lib/Server.js:219:19)
    at /Users/martin.svoboda/workspace/parcel-serve-fails-for-inlines/node_modules/@parcel/reporter-dev-server/lib/Server.js:441:14
    at call (/Users/martin.svoboda/workspace/parcel-serve-fails-for-inlines/node_modules/connect/index.js:239:7)
    at next (/Users/martin.svoboda/workspace/parcel-serve-fails-for-inlines/node_modules/connect/index.js:183:5)
    at Function.handle (/Users/martin.svoboda/workspace/parcel-serve-fails-for-inlines/node_modules/connect/index.js:186:3) {
  framesToPop: 1
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! parcel-serve-fails-for-inlines@ serve: `parcel serve ./src/index.html`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the parcel-serve-fails-for-inlines@ serve script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

```

## ✔️ PR Todo

- [x] Added/updated unit tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

martinsvoboda committed:
Fix dev server when serving inline bundles

martinsvoboda committed:
Merge branch 'v2' into serve-fails-for-inlines

mischnic commented:
The test would look similar to this one

https://github.com/parcel-bundler/parcel/blob/59e6f1c8d364045e75d7291419d3a16395f6f171/packages/core/integration-tests/test/server.js#L61-L85

where a index.html with
```html
<!DOCTYPE html>
<style>
body {
	background: red
}
</style>
```
is requested.
That should also trigger the error

martinsvoboda committed:
Extend html test for inline styles

martinsvoboda commented:
Thank you for the hint. Find the trigger for the error was tricky. Because "Error: Got unexpected null" was raised only when non-existing file was requested. Get default page scenario is actually covered by three tests (each is using integration/html/index.html template):
- should serve a default page if the main bundle is an HTML asset
- should serve a default page if the main bundle is an HTML asset even if it is not called index
- should serve a default page if the main bundle is an HTML asset with package.json#source

So I added inline style to integration/html/index.html and fixed missing inline bundles in other tests which also use this template.

wbinnssmith commented on packages/reporters/dev-server/src/Server.js:
```diff
@@ -207,6 +207,7 @@ export default class Server {
       let requestedPath = path.normalize(pathname.slice(1));
       let bundle = bundleGraph
         .getBundles()
```
cc @devongovett and @mischnic we recently discussed changing this method to never return inline bundles

martinsvoboda replied:
I tried to filter inline bundles in `getBundles` method, but many tests failed and I didn't figure out how to fix them. This change is definitely out of my knowledge of Parcel. However this bug is blocking my development, so should I finish this pull request as a quick fix, or you change `getBundles` method soon?

mischnic replied:
I think we should merge this for now (as there are also various other places where we already do `getBundles().filter(b => !b.isInline)` and change `getBundles` later.

mischnic commented on packages/core/integration-tests/test/server.js:
```diff
@@ -452,6 +452,11 @@ describe('server', function() {
         assets: ['index.html'],
       },
+      {
+        name: 'index.HASH_REF_2d285d6f29236f5b00e1db199488d505.css',
```
This hash isn't necessarily stable, it should work correctly even without the `name:` property.

martinsvoboda replied:
Perfect, I pushed the change. Probably there is a minor bug in assertBundles, because inline bundles have to be specified before index.html otherwise, assertBundles fails.

martinsvoboda committed:
Removed hash references from test bundle asserts (fixes unstable unit tests)

martinsvoboda commented:
Tests passed, the fix is ready for merge.
